File: src/main/java/com/wgzhao/addax/admin/service/impl/TargetServiceWithHiveImpl.java
package com.wgzhao.addax.admin.service.impl;

import com.wgzhao.addax.admin.common.JourKind;
import com.wgzhao.addax.admin.dto.HiveConnectDto;
import com.wgzhao.addax.admin.model.EtlJour;
import com.wgzhao.addax.admin.model.VwEtlTableWithSource;
import com.wgzhao.addax.admin.service.*;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.dbcp2.BasicDataSource;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import javax.sql.DataSource;

import java.io.File;
import java.net.MalformedURLException;
import java.net.URL;
import java.net.URLClassLoader;
import java.sql.Connection;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.List;
import java.text.MessageFormat;

@Service
@Slf4j
public class TargetServiceWithHiveImpl
        implements TargetService
{

    @Autowired
    private DictService dictService;

    @Autowired
    private ColumnService columnService;

    @Autowired
    private EtlJourService jourService;

    @Autowired
    private SystemConfigService configService;

    private volatile DataSource hiveDataSource;

    @Override
    public Connection getHiveConnect()
    {
        if (hiveDataSource == null) {
            synchronized (this) {
                if (hiveDataSource == null) {
                    HiveConnectDto hiveConnectDto = configService.getHiveServer2();
                    log.info("try to load hive jdbc driver from {}", hiveConnectDto.driverPath());
                    try {
                        hiveDataSource = getHiveDataSourceWithConfig(hiveConnectDto);
                        return hiveDataSource.getConnection();
                    }
                    catch (SQLException | MalformedURLException e) {
                        throw new RuntimeException(e);
                    }
                }
            }
        }
        try {
            return hiveDataSource.getConnection();
        }
        catch (Exception e) {
            throw new RuntimeException("Failed to get connection from Hive DataSource", e);
        }
    }

    @Override
    public DataSource getHiveDataSourceWithConfig(HiveConnectDto hiveConnectDto)
            throws MalformedURLException
    {

        File hiveJarFile = new File(hiveConnectDto.driverPath());
        URL[] jarUrls = new URL[] {hiveJarFile.toURI().toURL()};
        // 创建独立的类加载器
        URLClassLoader classLoader = new URLClassLoader(jarUrls, this.getClass().getClassLoader());

        // Set the context class loader
        Thread.currentThread().setContextClassLoader(classLoader);

        BasicDataSource dataSource = new BasicDataSource();
        dataSource.setUrl(hiveConnectDto.url());
        dataSource.setUsername(hiveConnectDto.username());
        dataSource.setPassword(hiveConnectDto.password());
        dataSource.setDriverClassName(hiveConnectDto.driverClassName());
        return dataSource;
    }

    /**
     * 为指定 Hive 表添加分区。
     *
     * @param taskId 采集任务ID
     * @param db Hive数据库名
     * @param table Hive表名
     * @param partName 分区字段名
     * @param partValue 分区字段值
     * @return 是否添加成功
     */
    @Override
    public boolean addPartition(long taskId, String db, String table, String partName, String partValue)
    {
        String sql = String.format("ALTER TABLE %s.%s ADD IF NOT EXISTS PARTITION (%s='%s')", db, table, partName, partValue);
        EtlJour etlJour = jourService.addJour(taskId, JourKind.PARTITION, sql);
        try (Connection conn = getHiveConnect();
                Statement stmt = conn.createStatement()) {
            log.info("Add partition for {}.{}: {}", db, table, sql);
            stmt.execute(sql);
            jourService.successJour(etlJour);
            return true;
        }
        catch (SQLException e) {
            log.error("Failed to add partition ", e);
            jourService.failJour(etlJour, e.getMessage());
            return false;
        }
    }

    /**
     * 创建或更新 Hive 目标表。
     * 包括建库、建表、分区、表属性等操作。
     *
     * @param etlTable 采集表视图对象
     * @return 是否创建/更新成功
     */
    @Override
    public boolean createOrUpdateHiveTable(VwEtlTableWithSource etlTable)
    {
        List<String> hiveColumns = columnService.getHiveColumnsAsDDL(etlTable.getId());

        String createDbSql = MessageFormat.format("create database if not exists `{0}` location ''{1}/{0}''",
                etlTable.getTargetDb(), dictService.getHdfsPrefix());
        String createTableSql = MessageFormat.format("""
                        create external table if not exists `{0}`.`{1}` (
                        {2}
                        ) comment ''{3}''
                        partitioned by ( `{4}` string )
                         stored as {5}
                         location ''{6}/{0}/{1}''
                         tblproperties (''external.table.purge''=''true'', ''discover.partitions''=''true'', ''orc.compress''=''{7}'', ''snappy.compress''=''{7}'')
                        """,
                etlTable.getTargetDb(), etlTable.getTargetTable(), String.join(",\n", hiveColumns), etlTable.getTblComment(),
                etlTable.getPartName(), dictService.getHdfsStorageFormat(), dictService.getHdfsPrefix(), dictService.getHdfsCompress());

        log.info("create table sql:\n{}", createTableSql);
        EtlJour etlJour = jourService.addJour(etlTable.getId(), JourKind.UPDATE_TABLE, createTableSql);
        try (Connection conn = getHiveConnect();
                Statement stmt = conn.createStatement()) {
            stmt.execute(createDbSql);
            stmt.execute(createTableSql);
            jourService.successJour(etlJour);
            return true;
        }
        catch (SQLException e) {
            log.warn("Failed to create or update hive table ", e);
            jourService.failJour(etlJour, e.getMessage());
            return false;
        }
    }
}

File: src/main/kotlin/com/wgzhao/addax/admin/service/impl/TargetServiceWithHiveImpl.kt
package com.wgzhao.addax.admin.service.impl

import com.wgzhao.addax.admin.common.JourKind
import com.wgzhao.addax.admin.model.VwEtlTableWithSource
import com.wgzhao.addax.admin.service.ColumnService
import com.wgzhao.addax.admin.service.DictService
import com.wgzhao.addax.admin.service.EtlJourService
import com.wgzhao.addax.admin.service.SystemConfigService
import com.wgzhao.addax.admin.service.TargetService
import io.github.oshai.kotlinlogging.KotlinLogging
import org.apache.commons.dbcp2.BasicDataSource
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import java.io.File
import java.net.URLClassLoader
import java.sql.Connection
import java.sql.SQLException
import javax.sql.DataSource
import kotlin.concurrent.Volatile

@Service
class TargetServiceWithHiveImpl(
    private val dictService: DictService,
    private val columnService: ColumnService,
    private val jourService: EtlJourService,
    private val configService: SystemConfigService,
) : TargetService {

    private val logger = KotlinLogging.logger {}


    @Value("\${spring.datasource.hive.url}")
    lateinit var  url: String

    @Value("\${spring.datasource.hive.username}")
    private val username: String? = null

    @Value("\${spring.datasource.hive.password}")
    private val password: String? = null

    @Value("\${spring.datasource.hive.driver-class-name}")
    private val driverClassName: String? = null

    @Value("\${spring.datasource.hive.jar-path}")
    lateinit var driverPath: String

    @Volatile
    private var hiveDataSource: DataSource? = null

    private fun getHiveDataSource(): Connection {
        if (hiveDataSource == null) {
            synchronized(this) {
                if (hiveDataSource == null) {
                    try {
                        logger.info { "try to load hive jdbc driver from $driverPath" }
                        val hiveJarFile = File(driverPath)
                        val jarUrls = arrayOf(hiveJarFile.toURI().toURL())
                        val classLoader = URLClassLoader(jarUrls, this.javaClass.classLoader)
                        Thread.currentThread().contextClassLoader = classLoader
                        val dataSource = BasicDataSource()
                        dataSource.url = url
                        dataSource.username = username
                        dataSource.password = password
                        dataSource.driverClassName = driverClassName
                        hiveDataSource = dataSource
                    } catch (e: Exception) {
                        throw RuntimeException(e)
                    }
                }
            }
        }
        try {
            return hiveDataSource!!.connection
        } catch (e: Exception) {
            throw RuntimeException("Failed to get connection from Hive DataSource", e)
        }
    }

    override fun addPartition(taskId: Long, db: String, table: String, partName: String, partValue: String): Boolean {
        val sql = "ALTER TABLE $db.$table ADD IF NOT EXISTS PARTITION ($partName='$partValue')"
        val etlJour = jourService.addJour(taskId, JourKind.PARTITION, sql)
        try {
            getHiveDataSource().use { conn ->
                conn.createStatement().use { stmt ->
                    logger.info { "Add partition for $db.$table: $sql" }
                    stmt.execute(sql)
                    jourService.successJour(etlJour)
                    return true
                }
            }
        } catch (e: SQLException) {
            logger.error(e) { "Failed to add partition" }
            jourService.failJour(etlJour, e.message)
            return false
        }
    }

    override fun createOrUpdateHiveTable(etlTable: VwEtlTableWithSource): Boolean {
        val hiveColumns = columnService.getHiveColumnsAsDDL(etlTable.id)
        val createDbSql = "create database if not exists `${etlTable.targetDb}` location '${configService.hdfsPrefix}/${etlTable.targetDb}'"
        val createTableSql = """
            create external table if not exists `${etlTable.targetDb}`.`${etlTable.targetTable}` (
            ${hiveColumns.joinToString(",\n")}
            ) comment '${etlTable.tblComment}'
            partitioned by ( `${etlTable.partName}` string )
            stored as ${etlTable.storageFormat}
            location '${configService.hdfsPrefix}/${etlTable.targetDb}/${etlTable.targetTable}'
            tblproperties ('external.table.purge'='true', 'discover.partitions'='true', 'orc.compress'='${etlTable.compressFormat}', 'snappy.compress'='${etlTable.compressFormat}')
        """.trimIndent()
        logger.info { "create table sql:\n$createTableSql" }
        val etlJour = jourService.addJour(etlTable.id, JourKind.UPDATE_TABLE, createTableSql)
        try {
            getHiveDataSource().use { conn ->
                conn.createStatement().use { stmt ->
                    stmt.execute(createDbSql)
                    stmt.execute(createTableSql)
                    jourService.successJour(etlJour)
                    return true
                }
            }
        } catch (e: SQLException) {
            logger.warn(e) { "Failed to create or update hive table" }
            jourService.failJour(etlJour, e.message)
            return false
        }
    }
}
